File: app/src/main/java/com/jkm/popularmovies/DetailActivity.java
package com.jkm.popularmovies;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Intent;
import android.content.res.Configuration;
import android.database.Cursor;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.jkm.popularmovies.data.MovieContract;
import com.jkm.popularmovies.model.MovieModel;
import com.jkm.popularmovies.model.ReviewModel;
import com.jkm.popularmovies.model.ReviewResultModel;
import com.jkm.popularmovies.model.TrailerModel;
import com.jkm.popularmovies.model.TrailerResultModel;
import com.jkm.popularmovies.util.EndlessRecyclerViewScrollListener;
import com.squareup.picasso.Picasso;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public class DetailActivity extends AppCompatActivity {
    private static final String TAG = DetailActivity.class.getSimpleName();

    @BindView(R.id.tv_name_detail)
    TextView nameTextView;
    @BindView(R.id.iv_poster_detail)
    ImageView posterImageView;
    @BindView(R.id.tv_release_date_detail)
    TextView releaseDateTextView;
    @BindView(R.id.tv_rating_detail)
    TextView ratingTextView;
    @BindView(R.id.bt_add_favorite)
    Button addFavoriteButton;
    @BindView(R.id.tv_overview_detail)
    TextView overviewTextView;
    @BindView(R.id.layout_movie_part_detail)
    LinearLayout moviePartDetailLayout;

    @BindView(R.id.collapsing_toolbar_detail)
    CollapsingToolbarLayout collapsingToolbar;
    @BindView(R.id.tab_trailer_detail)
    TabLayout trailerTabLayout;
    @BindView(R.id.view_pager_trailer_detail)
    ViewPager trailerViewPager;
    @BindView(R.id.pb_trailer_detail)
    ProgressBar trailerProgressBar;
    @BindView(R.id.toolbar_detail)
    Toolbar toolbar;

    @BindView(R.id.rv_review_detail)
    RecyclerView reviewRecyclerView;
    @BindView(R.id.tv_review_empty)
    TextView reviewEmptyTextView;
    @BindView(R.id.pb_review_detail)
    ProgressBar reviewProgressBar;

    private ReviewAdapter mReviewAdapter;
    private MovieModel mMovieModel;
    private ArrayList<ReviewResultModel> mReviewResultModels = new ArrayList<>();
    private ArrayList<TrailerResultModel> mTrailerResultModels = new ArrayList<>();

    private DetailApiInterface mDetailApiInterface;

    private int reviewPage = 2;
    private static int totalReviewPages;
    private boolean isFavorite = false;

    private EndlessRecyclerViewScrollListener mScrollListener;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail);
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        int orientation = getResources().getConfiguration().orientation;
        setToolbarViewParams(orientation);
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            setLinearLayoutWeight(moviePartDetailLayout, 2);
        } else {
            setLinearLayoutWeight(moviePartDetailLayout, 1);
        }

        Retrofit retrofit = new Retrofit.Builder().baseUrl(BuildConfig.MOVIE_DB_API_URL)
                .addConverterFactory(GsonConverterFactory.create()).build();
        mDetailApiInterface = retrofit.create(DetailApiInterface.class);

        Intent intent = getIntent();
        if (intent.hasExtra(getString(R.string.detail_key))) {
            mMovieModel = intent.getParcelableExtra(getString(R.string.detail_key));

            isFavorite = isFavoriteMovie();
            renderView(mMovieModel);
            configureRecyclerView(reviewRecyclerView);

            if (savedInstanceState != null) {
                mTrailerResultModels = savedInstanceState.getParcelableArrayList(getString(R.string.trailer_key));
                configureViewPagerWithTabLayout(mTrailerResultModels);

                mReviewResultModels = savedInstanceState.getParcelableArrayList(getString(R.string.review_key));
                reviewPage = savedInstanceState.getInt(getString(R.string.review_page_key));
            } else {
                callTrailers(mMovieModel.getId());
                callReviews(mMovieModel.getId(), 1, true);
            }

            mReviewAdapter = new ReviewAdapter(DetailActivity.this, mReviewResultModels);
            reviewRecyclerView.setAdapter(mReviewAdapter);
        } else {
            Toast.makeText(this, getString(R.string.failed_show_detail), Toast.LENGTH_LONG).show();
        }

        addFavoriteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (isFavorite) {
                    deleteFromDatabase();
                    Toast.makeText(DetailActivity.this, getString(R.string.removed_from_favorites), Toast.LENGTH_SHORT).show();
                    addFavoriteButton.setText(getString(R.string.add_favorite));
                } else {
                    insertToDatabase();
                    Toast.makeText(DetailActivity.this, getString(R.string.added_to_favorites), Toast.LENGTH_SHORT).show();
                    addFavoriteButton.setText(getString(R.string.remove_favorite));
                }
            }
        });
    }

    private boolean isFavoriteMovie() {
        ContentResolver detailContentResolver = getContentResolver();
        Cursor cursor = detailContentResolver.query(
                MovieContract.MovieEntry.buildMovieUriWithMovieId(mMovieModel.getId()),
                null,
                null,
                null,
                MovieContract.MovieEntry._ID + " ASC");

        if (cursor != null) {
            cursor.close();
            return cursor.getCount() > 0;
        } else {
            return false;
        }
    }

    private void insertToDatabase() {
        ContentResolver detailContentResolver = getContentResolver();

        ContentValues movieValues = new ContentValues();
        movieValues.put(MovieContract.MovieEntry.COLUMN_MOVIE_ID, mMovieModel.getId());
        movieValues.put(MovieContract.MovieEntry.COLUMN_MOVIE_NAME, mMovieModel.getOriginalTitle());
        movieValues.put(MovieContract.MovieEntry.COLUMN_POSTER_PATH, mMovieModel.getPosterPath());
        movieValues.put(MovieContract.MovieEntry.COLUMN_OVERVIEW, mMovieModel.getOverview());
        movieValues.put(MovieContract.MovieEntry.COLUMN_USER_RATING, mMovieModel.getVoteAverage());
        movieValues.put(MovieContract.MovieEntry.COLUMN_RELEASE_DATE, mMovieModel.getReleaseDate());

        detailContentResolver.insert(MovieContract.MovieEntry.CONTENT_URI, movieValues);
    }

    private void deleteFromDatabase() {
        ContentResolver detailContentResolver = getContentResolver();
        String[] selectionArguments = new String[]{String.valueOf(mMovieModel.getId())};
        detailContentResolver.delete(MovieContract.MovieEntry.CONTENT_URI,
                MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ", selectionArguments);
    }

    private void renderView(MovieModel movieModel) {
        collapsingToolbar.setTitle(movieModel.getOriginalTitle());
        collapsingToolbar.setCollapsedTitleTextColor(Color.WHITE);
        collapsingToolbar.setExpandedTitleColor(Color.TRANSPARENT);

        nameTextView.setText(movieModel.getOriginalTitle());

        String posterPath = BuildConfig.MOVIE_DB_POSTER_URL + movieModel.getPosterPath();
        Picasso.with(this)
                .load(posterPath)
                .placeholder(R.drawable.placeholder)
                .error(R.drawable.placeholder)
                .into(posterImageView);

        String releaseDate = getFormattedDate(movieModel.getReleaseDate(), getString(R.string.movie_db_date_format),
                getString(R.string.date_format));
        releaseDateTextView.setText(releaseDate);

        String rating = String.valueOf(movieModel.getVoteAverage()) + getString(R.string.max_rating);
        ratingTextView.setText(rating);

        if (isFavorite) addFavoriteButton.setText(getString(R.string.remove_favorite));
        else addFavoriteButton.setText(getString(R.string.add_favorite));

        overviewTextView.setText(movieModel.getOverview());
    }

    private void configureRecyclerView(RecyclerView recyclerView) {
        LinearLayoutManager layoutManager = new LinearLayoutManager(DetailActivity.this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setNestedScrollingEnabled(false);

        mScrollListener = new EndlessRecyclerViewScrollListener(layoutManager) {
            @Override
            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
                if (reviewPage <= totalReviewPages) {
                    callReviews(mMovieModel.getId(), reviewPage, false);
                }
            }
        };
        recyclerView.addOnScrollListener(mScrollListener);
    }

    private void configureViewPagerWithTabLayout(ArrayList<TrailerResultModel> results) {
        trailerViewPager.setAdapter(new TrailerViewPagerAdapter(getSupportFragmentManager(), results));
        trailerTabLayout.setupWithViewPager(trailerViewPager, true);
    }

    private void callTrailers(int id) {
        showTrailerLoading();

        Call<TrailerModel> trailers = mDetailApiInterface.getTrailers(id, BuildConfig.MOVIE_DB_API_KEY);
        trailers.enqueue(new Callback<TrailerModel>() {
            @Override
            public void onResponse(Call<TrailerModel> call, Response<TrailerModel> response) {
                if (response.body() != null) {
                    ArrayList<TrailerResultModel> results = response.body().getResults();
                    if (results != null && !results.isEmpty()) {
                        mTrailerResultModels.clear();

                        for (TrailerResultModel result : results) {
                            if (result.getSite().equals(getString(R.string.youtube))) {
                                mTrailerResultModels.add(result);
                            }
                        }

                        configureViewPagerWithTabLayout(mTrailerResultModels);
                        hideTrailerLoading();
                    } else {
                        hideAllTrailerLayout();
                    }
                } else {
                    Toast.makeText(DetailActivity.this, R.string.failed_fetch_trailers, Toast.LENGTH_SHORT).show();
                    hideAllTrailerLayout();
                }
            }

            @Override
            public void onFailure(Call<TrailerModel> call, Throwable t) {
                Log.e(TAG, getString(R.string.failed_fetch_trailers), t.getCause());
                Toast.makeText(DetailActivity.this, R.string.failed_fetch_trailers, Toast.LENGTH_SHORT).show();
                hideAllTrailerLayout();
            }
        });
    }

    private void callReviews(int id, int page, final boolean clearList) {
        if (clearList) showReviewLoading();

        Call<ReviewModel> reviews = mDetailApiInterface.getReviews(id, BuildConfig.MOVIE_DB_API_KEY, page);
        reviews.enqueue(new Callback<ReviewModel>() {
            @Override
            public void onResponse(Call<ReviewModel> call, Response<ReviewModel> response) {
                if (response.body() != null) {
                    totalReviewPages = response.body().getTotalPages();
                    ArrayList<ReviewResultModel> results = response.body().getResults();
                    if (results != null && !results.isEmpty()) {
                        if (clearList) mReviewResultModels.clear();
                        else reviewPage++;

                        mReviewResultModels.addAll(results);
                        mReviewAdapter.notifyDataSetChanged();

                        mScrollListener.resetState();
                        hideReviewLoading();
                    } else {
                        showReviewEmptyMessage();
                    }
                } else {
                    Toast.makeText(DetailActivity.this, R.string.failed_fetch_reviews, Toast.LENGTH_SHORT).show();
                    showReviewEmptyMessage();
                }
            }

            @Override
            public void onFailure(Call<ReviewModel> call, Throwable t) {
                Log.e(TAG, getString(R.string.failed_fetch_reviews), t.getCause());
                Toast.makeText(DetailActivity.this, R.string.failed_fetch_reviews, Toast.LENGTH_SHORT).show();
                hideReviewLoading();
            }
        });
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelableArrayList(getString(R.string.trailer_key), mTrailerResultModels);
        outState.putParcelableArrayList(getString(R.string.review_key), mReviewResultModels);
        outState.putInt(getString(R.string.review_page_key), reviewPage);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    private String getFormattedDate(String date, String currentFormat, String targetFormat) {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(currentFormat, Locale.ENGLISH);
        try {
            Date tempDate = simpleDateFormat.parse(date);
            SimpleDateFormat dateFormat = new SimpleDateFormat(targetFormat, Locale.ENGLISH);
            return dateFormat.format(tempDate);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return null;
    }

    private void showTrailerLoading() {
        trailerProgressBar.setVisibility(View.VISIBLE);
        trailerViewPager.setVisibility(View.INVISIBLE);
        trailerTabLayout.setVisibility(View.INVISIBLE);
    }

    private void hideTrailerLoading() {
        trailerProgressBar.setVisibility(View.GONE);
        trailerViewPager.setVisibility(View.VISIBLE);
        trailerTabLayout.setVisibility(View.VISIBLE);
    }

    private void hideAllTrailerLayout() {
        trailerProgressBar.setVisibility(View.GONE);
        trailerViewPager.setVisibility(View.INVISIBLE);
        trailerTabLayout.setVisibility(View.INVISIBLE);
    }

    private void showReviewLoading() {
        reviewRecyclerView.setVisibility(View.INVISIBLE);
        reviewProgressBar.setVisibility(View.VISIBLE);
        reviewEmptyTextView.setVisibility(View.GONE);
    }

    private void hideReviewLoading() {
        reviewRecyclerView.setVisibility(View.VISIBLE);
        reviewProgressBar.setVisibility(View.GONE);
        reviewEmptyTextView.setVisibility(View.GONE);
    }

    private void showReviewEmptyMessage() {
        reviewRecyclerView.setVisibility(View.INVISIBLE);
        reviewProgressBar.setVisibility(View.GONE);
        reviewEmptyTextView.setVisibility(View.VISIBLE);
    }

    private void setLinearLayoutWeight(LinearLayout linearLayout, float weight) {
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT);
        layoutParams.weight = weight;
        linearLayout.setLayoutParams(layoutParams);
    }

    private void setToolbarViewParams(int orientation) {
        DisplayMetrics displaymetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displaymetrics);
        int width = displaymetrics.widthPixels;
        int height = displaymetrics.heightPixels;

        int viewPagerHeight, tabLayoutMargin;
        if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            viewPagerHeight = (width * 4) / 9;
        } else {
            viewPagerHeight = (height * 4 / 9);
        }
        tabLayoutMargin = width / 3;

        CollapsingToolbarLayout.LayoutParams layoutParamsForViewPager = new CollapsingToolbarLayout
                .LayoutParams(CollapsingToolbarLayout.LayoutParams.MATCH_PARENT, viewPagerHeight);
        trailerViewPager.setLayoutParams(layoutParamsForViewPager);

        CollapsingToolbarLayout.LayoutParams layoutParamsForTabLayout = new CollapsingToolbarLayout
                .LayoutParams(CollapsingToolbarLayout.LayoutParams.WRAP_CONTENT, CollapsingToolbarLayout.LayoutParams.WRAP_CONTENT);
        layoutParamsForTabLayout.rightMargin = tabLayoutMargin;
        layoutParamsForTabLayout.leftMargin = tabLayoutMargin;
        layoutParamsForTabLayout.gravity = Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
        trailerTabLayout.setLayoutParams(layoutParamsForTabLayout);
    }

    interface DetailApiInterface {
        @GET("movie/{id}/reviews")
        Call<ReviewModel> getReviews(@Path("id") int id, @Query("api_key") String apiKey, @Query("page") int page);

        @GET("movie/{id}/videos")
        Call<TrailerModel> getTrailers(@Path("id") int id, @Query("api_key") String apiKey);
    }

    private class TrailerViewPagerAdapter extends FragmentPagerAdapter {
        private ArrayList<TrailerResultModel> results;

        TrailerViewPagerAdapter(FragmentManager manager, ArrayList<TrailerResultModel> results) {
            super(manager);
            this.results = results;
        }

        @Override
        public Fragment getItem(int position) {
            return TrailerFragment.newInstance(DetailActivity.this, results.get(position).getKey());
        }

        @Override
        public int getCount() {
            if (results == null) return 0;
            else return results.size();
        }
    }
}

File: app/src/main/java/com/jkm/popularmovies/DetailActivity.kt
package com.jkm.popularmovies

import android.content.ContentValues
import android.content.res.Configuration
import android.graphics.Color
import android.os.Bundle
import com.google.android.material.appbar.CollapsingToolbarLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.DisplayMetrics
import android.util.Log
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import com.jkm.popularmovies.data.MovieContract
import com.jkm.popularmovies.model.*
import com.jkm.popularmovies.util.EndlessRecyclerViewScrollListener
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_detail.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class DetailActivity : AppCompatActivity() {

    private var mReviewAdapter: ReviewAdapter? = null
    private var mMovieModel: MovieModel? = null
    private var mReviewResultModels: ArrayList<ReviewResultModel>? = ArrayList()
    private var mTrailerResultModels: ArrayList<TrailerResultModel>? = ArrayList()

    private var mDetailApiInterface: DetailApiInterface? = null

    private var reviewPage = 2
    private var isFavorite = false

    private var mScrollListener: EndlessRecyclerViewScrollListener? = null

    private val isFavoriteMovie: Boolean
        get() {
            val detailContentResolver = contentResolver
            val cursor = detailContentResolver.query(
                    MovieContract.MovieEntry.buildMovieUriWithMovieId(mMovieModel!!.id), null, null, null,
                    MovieContract.MovieEntry._ID + " ASC")

            if (cursor != null) {
                cursor.close()
                return cursor.count > 0
            } else {
                return false
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val orientation = resources.configuration.orientation
        setToolbarViewParams(orientation)
        if (orientation == Configuration.ORIENTATION_LANDSCAPE) {
            setLinearLayoutWeight(moviePartDetailLayout!!, 2f)
        } else {
            setLinearLayoutWeight(moviePartDetailLayout!!, 1f)
        }

        val retrofit = Retrofit.Builder().baseUrl(BuildConfig.MOVIE_DB_API_URL)
                .addConverterFactory(GsonConverterFactory.create()).build()
        mDetailApiInterface = retrofit.create(DetailApiInterface::class.java)

        val intent = intent
        if (intent.hasExtra(getString(R.string.detail_key))) {
            mMovieModel = intent.getParcelableExtra(getString(R.string.detail_key))

            isFavorite = isFavoriteMovie
            renderView(mMovieModel!!)
            configureRecyclerView(reviewRecyclerView!!)

            if (savedInstanceState != null) {
                mTrailerResultModels = savedInstanceState.getParcelableArrayList(getString(R.string.trailer_key))
                configureViewPagerWithTabLayout(mTrailerResultModels)

                mReviewResultModels = savedInstanceState.getParcelableArrayList(getString(R.string.review_key))
                reviewPage = savedInstanceState.getInt(getString(R.string.review_page_key))
            } else {
                callTrailers(mMovieModel!!.id)
                callReviews(mMovieModel!!.id, 1, true)
            }

            mReviewAdapter = ReviewAdapter(this@DetailActivity, mReviewResultModels)
            reviewRecyclerView!!.adapter = mReviewAdapter
        } else {
            Toast.makeText(this, getString(R.string.failed_show_detail), Toast.LENGTH_LONG).show()
        }

        addFavoriteButton!!.setOnClickListener {
            if (isFavorite) {
                deleteFromDatabase()
                Toast.makeText(this@DetailActivity, getString(R.string.removed_from_favorites), Toast.LENGTH_SHORT).show()
                addFavoriteButton!!.text = getString(R.string.add_favorite)
            } else {
                insertToDatabase()
                Toast.makeText(this@DetailActivity, getString(R.string.added_to_favorites), Toast.LENGTH_SHORT).show()
                addFavoriteButton!!.text = getString(R.string.remove_favorite)
            }
        }
    }

    private fun insertToDatabase() {
        val detailContentResolver = contentResolver

        val movieValues = ContentValues()
        movieValues.put(MovieContract.MovieEntry.COLUMN_MOVIE_ID, mMovieModel!!.id)
        movieValues.put(MovieContract.MovieEntry.COLUMN_MOVIE_NAME, mMovieModel!!.originalTitle)
        movieValues.put(MovieContract.MovieEntry.COLUMN_POSTER_PATH, mMovieModel!!.posterPath)
        movieValues.put(MovieContract.MovieEntry.COLUMN_OVERVIEW, mMovieModel!!.overview)
        movieValues.put(MovieContract.MovieEntry.COLUMN_USER_RATING, mMovieModel!!.voteAverage)
        movieValues.put(MovieContract.MovieEntry.COLUMN_RELEASE_DATE, mMovieModel!!.releaseDate)

        detailContentResolver.insert(MovieContract.MovieEntry.CONTENT_URI, movieValues)
    }

    private fun deleteFromDatabase() {
        val detailContentResolver = contentResolver
        val selectionArguments = arrayOf(mMovieModel!!.id.toString())
        detailContentResolver.delete(MovieContract.MovieEntry.CONTENT_URI,
                MovieContract.MovieEntry.COLUMN_MOVIE_ID + " = ? ", selectionArguments)
    }

    private fun renderView(movieModel: MovieModel) {
        collapsingToolbar!!.title = movieModel.originalTitle
        collapsingToolbar!!.setCollapsedTitleTextColor(Color.WHITE)
        collapsingToolbar!!.setExpandedTitleColor(Color.TRANSPARENT)

        nameTextView!!.text = movieModel.originalTitle

        val posterPath = BuildConfig.MOVIE_DB_POSTER_URL + movieModel.posterPath
        Picasso.with(this)
                .load(posterPath)
                .placeholder(R.drawable.placeholder)
                .error(R.drawable.placeholder)
                .into(posterImageView)

        val releaseDate = getFormattedDate(movieModel.releaseDate, getString(R.string.movie_db_date_format),
                getString(R.string.date_format))
        releaseDateTextView!!.text = releaseDate

        val rating = movieModel.voteAverage.toString() + getString(R.string.max_rating)
        ratingTextView!!.text = rating

        if (isFavorite)
            addFavoriteButton!!.text = getString(R.string.remove_favorite)
        else
            addFavoriteButton!!.text = getString(R.string.add_favorite)

        overviewTextView!!.text = movieModel.overview
    }

    private fun configureRecyclerView(recyclerView: RecyclerView) {
        val layoutManager = LinearLayoutManager(this@DetailActivity)
        recyclerView.layoutManager = layoutManager
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.isNestedScrollingEnabled = false

        mScrollListener = object : EndlessRecyclerViewScrollListener(layoutManager) {
            override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView) {
                if (reviewPage <= totalReviewPages) {
                    callReviews(mMovieModel!!.id, reviewPage, false)
                }
            }
        }
        recyclerView.addOnScrollListener(mScrollListener!!)
    }

    private fun configureViewPagerWithTabLayout(results: ArrayList<TrailerResultModel>?) {
        trailerViewPager!!.adapter = TrailerViewPagerAdapter(supportFragmentManager, results)
        trailerTabLayout!!.setupWithViewPager(trailerViewPager, true)
    }

    private fun callTrailers(id: Int) {
        showTrailerLoading()

        val trailers = mDetailApiInterface!!.getTrailers(id, BuildConfig.MOVIE_DB_API_KEY)
        trailers.enqueue(object : Callback<TrailerModel> {
            override fun onResponse(call: Call<TrailerModel>, response: Response<TrailerModel>) {
                if (response.body() != null) {
                    val results = response.body()!!.results
                    if (results != null && !results.isEmpty()) {
                        mTrailerResultModels!!.clear()

                        for (result in results) {
                            if (result.site == getString(R.string.youtube)) {
                                mTrailerResultModels!!.add(result)
                            }
                        }

                        configureViewPagerWithTabLayout(mTrailerResultModels)
                        hideTrailerLoading()
                    } else {
                        hideAllTrailerLayout()
                    }
                } else {
                    Toast.makeText(this@DetailActivity, R.string.failed_fetch_trailers, Toast.LENGTH_SHORT).show()
                    hideAllTrailerLayout()
                }
            }

            override fun onFailure(call: Call<TrailerModel>, t: Throwable) {
                Log.e(TAG, getString(R.string.failed_fetch_trailers), t.cause)
                Toast.makeText(this@DetailActivity, R.string.failed_fetch_trailers, Toast.LENGTH_SHORT).show()
                hideAllTrailerLayout()
            }
        })
    }

    private fun callReviews(id: Int, page: Int, clearList: Boolean) {
        if (clearList) showReviewLoading()

        val reviews = mDetailApiInterface!!.getReviews(id, BuildConfig.MOVIE_DB_API_KEY, page)
        reviews.enqueue(object : Callback<ReviewModel> {
            override fun onResponse(call: Call<ReviewModel>, response: Response<ReviewModel>) {
                if (response.body() != null) {
                    totalReviewPages = response.body()!!.totalPages
                    val results = response.body()!!.results
                    if (results != null && !results.isEmpty()) {
                        if (clearList)
                            mReviewResultModels!!.clear()
                        else
                            reviewPage++

                        mReviewResultModels!!.addAll(results)
                        mReviewAdapter!!.notifyDataSetChanged()

                        mScrollListener!!.resetState()
                        hideReviewLoading()
                    } else {
                        showReviewEmptyMessage()
                    }
                } else {
                    Toast.makeText(this@DetailActivity, R.string.failed_fetch_reviews, Toast.LENGTH_SHORT).show()
                    showReviewEmptyMessage()
                }
            }

            override fun onFailure(call: Call<ReviewModel>, t: Throwable) {
                Log.e(TAG, getString(R.string.failed_fetch_reviews), t.cause)
                Toast.makeText(this@DetailActivity, R.string.failed_fetch_reviews, Toast.LENGTH_SHORT).show()
                hideReviewLoading()
            }
        })
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(getString(R.string.trailer_key), mTrailerResultModels)
        outState.putParcelableArrayList(getString(R.string.review_key), mReviewResultModels)
        outState.putInt(getString(R.string.review_page_key), reviewPage)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun getFormattedDate(date: String, currentFormat: String, targetFormat: String): String? {
        val simpleDateFormat = SimpleDateFormat(currentFormat, Locale.ENGLISH)
        try {
            val tempDate = simpleDateFormat.parse(date)
            val dateFormat = SimpleDateFormat(targetFormat, Locale.ENGLISH)
            return dateFormat.format(tempDate)
        } catch (e: ParseException) {
            e.printStackTrace()
        }

        return null
    }

    private fun showTrailerLoading() {
        trailerProgressBar!!.visibility = View.VISIBLE
        trailerViewPager!!.visibility = View.INVISIBLE
        trailerTabLayout!!.visibility = View.INVISIBLE
    }

    private fun hideTrailerLoading() {
        trailerProgressBar!!.visibility = View.GONE
        trailerViewPager!!.visibility = View.VISIBLE
        trailerTabLayout!!.visibility = View.VISIBLE
    }

    private fun hideAllTrailerLayout() {
        trailerProgressBar!!.visibility = View.GONE
        trailerViewPager!!.visibility = View.INVISIBLE
        trailerTabLayout!!.visibility = View.INVISIBLE
    }

    private fun showReviewLoading() {
        reviewRecyclerView!!.visibility = View.INVISIBLE
        reviewProgressBar!!.visibility = View.VISIBLE
        reviewEmptyTextView!!.visibility = View.GONE
    }

    private fun hideReviewLoading() {
        reviewRecyclerView!!.visibility = View.VISIBLE
        reviewProgressBar!!.visibility = View.GONE
        reviewEmptyTextView!!.visibility = View.GONE
    }

    private fun showReviewEmptyMessage() {
        reviewRecyclerView!!.visibility = View.INVISIBLE
        reviewProgressBar!!.visibility = View.GONE
        reviewEmptyTextView!!.visibility = View.VISIBLE
    }

    private fun setLinearLayoutWeight(linearLayout: LinearLayout, weight: Float) {
        val layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT)
        layoutParams.weight = weight
        linearLayout.layoutParams = layoutParams
    }

    private fun setToolbarViewParams(orientation: Int) {
        val displaymetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displaymetrics)
        val width = displaymetrics.widthPixels
        val height = displaymetrics.heightPixels

        val viewPagerHeight: Int
        val tabLayoutMargin: Int
        if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            viewPagerHeight = width * 4 / 9
        } else {
            viewPagerHeight = height * 4 / 9
        }
        tabLayoutMargin = width / 3

        val layoutParamsForViewPager = CollapsingToolbarLayout.LayoutParams(CollapsingToolbarLayout.LayoutParams.MATCH_PARENT, viewPagerHeight)
        trailerViewPager!!.layoutParams = layoutParamsForViewPager

        val layoutParamsForTabLayout = CollapsingToolbarLayout.LayoutParams(CollapsingToolbarLayout.LayoutParams.WRAP_CONTENT, CollapsingToolbarLayout.LayoutParams.WRAP_CONTENT)
        layoutParamsForTabLayout.rightMargin = tabLayoutMargin
        layoutParamsForTabLayout.leftMargin = tabLayoutMargin
        layoutParamsForTabLayout.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
        trailerTabLayout!!.layoutParams = layoutParamsForTabLayout
    }

    internal interface DetailApiInterface {
        @GET("movie/{id}/reviews")
        fun getReviews(@Path("id") id: Int, @Query("api_key") apiKey: String, @Query("page") page: Int): Call<ReviewModel>

        @GET("movie/{id}/videos")
        fun getTrailers(@Path("id") id: Int, @Query("api_key") apiKey: String): Call<TrailerModel>
    }

    private inner class TrailerViewPagerAdapter internal constructor(manager: FragmentManager, private val results: ArrayList<TrailerResultModel>?) : FragmentPagerAdapter(manager) {

        override fun getItem(position: Int): Fragment {
            return TrailerFragment.newInstance(this@DetailActivity, results!![position].key)
        }

        override fun getCount(): Int {
            return results?.size ?: 0
        }
    }

    companion object {
        private val TAG = DetailActivity::class.java.simpleName
        private var totalReviewPages: Int = 0
    }
}